### PR TITLE
add MVP with Cafe classes and order logic

### PR DESCRIPTION
**Pull request by tudorsonycx**

### CODE REVIEW

File: src/main/scala/Bill.scala [hunk 1]
```diff
@@ -0,0 +1,14 @@
+case class Bill(customer: Customer, items: Map[Item, Int], serviceCharge: Double) {
+  def total: Double = {
+    val itemTotal = items.map({
+      case (item, quantity) => item.price * quantity
```
Comment (eloiseyaz) [inline src/main/scala/Bill.scala:4]: Why do we need the case here and on line 10?

File: src/main/scala/Bill.scala [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    }).sum
+    itemTotal
+  }
+
+  override def toString: String = {
```
Comment (eloiseyaz) [inline src/main/scala/Bill.scala:11]: Overriding toString is a cool idea.

File: src/main/scala/Menu.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import scala.annotation.tailrec
+
+case class Menu private(items: List[Item]) {
```
Comment (eloiseyaz) [inline src/main/scala/Menu.scala:3]: What does the private do here?

File: src/test/scala/CafeSpec.scala [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+  val cafe: Cafe = Cafe("Test Cafe", Menu(List(item0, item1, item2, item3)))
+
+  "Cafe.showMenu" should {
+    "return a string representation of the menu" in {
+      val expectedResult: String =
+        "Test Item 0 (ColdDrink) - 5.0\nTest Item 1 (ColdFood) - 10.0\nTest Item 2 (HotFood) - 20.0\nTest Item 3 (PremiumMeal) - 30.0"
+
+      cafe.showMenu shouldBe expectedResult
+    }
```
Comment (eloiseyaz) [inline src/test/scala/CafeSpec.scala:25]: Add in testing for placeOrder - seems like at least 8 cases but may be more depending on edge cases.

File: src/main/scala/Item.scala [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+case object HotDrink extends ItemCategory
+
+case object ColdDrink extends ItemCategory
+
+case object AlcoholicDrink extends ItemCategory
+
+case object PremiumMeal extends ItemCategory
+
+case class Item(name: String, price: Double, category: ItemCategory, stock: Int = 0) {
+  override def toString: String = s"$name ($category) - $price"
```
Comment (eloiseyaz) [inline src/main/scala/Item.scala:16]: I like the toString override - makes items more readable.
Would be good to add some short testing though.

File: src/main/scala/Bill.scala (hunk 2 above)
Comment (eloiseyaz) [reply src/main/scala/Bill.scala:11]: Can add some short tests for this method.

File: src/main/scala/Bill.scala [hunk 6]
```diff
@@ -0,0 +1,14 @@
+case class Bill(customer: Customer, items: Map[Item, Int], serviceCharge: Double) {
+  def total: Double = {
```
Comment (eloiseyaz) [inline src/main/scala/Bill.scala]: Add some tests for this method.

File: src/test/scala/CafeSpec.scala
Comment (eloiseyaz) [inline src/test/scala/CafeSpec.scala:1]: You mentioned problems with your tests but they all pass so far 👍

File: src/test/scala/MenuSpec.scala [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  val item1: Item = Item("Test Item 1", 10.0, ColdFood, 5)
+  val item2: Item = Item("Test Item 2", 20.0, HotFood, 5)
+  val item3: Item = Item("Test Item 3", 30.0, PremiumMeal, 3)
+
+  val itemInvalid: Item = Item("Invalid Item", 10.0, ColdFood)
+
+  val menu: Menu = Menu(List(item0, item1, item2, item3))
+
+  "Menu.toString" should {
+    "return a string representation of the menu" in {
```
Comment (eloiseyaz) [inline src/test/scala/MenuSpec.scala:13]: Can add in edge case of an empty menu

File: src/test/scala/MenuSpec.scala [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+      menu.toString shouldBe "Test Item 0 (ColdDrink) - 5.0\nTest Item 1 (ColdFood) - 10.0\nTest Item 2 (HotFood) - 20.0\nTest Item 3 (PremiumMeal) - 30.0"
+    }
+  }
+
+  "Menu.updateMenuItem" should {
```
Comment (eloiseyaz) [inline src/test/scala/MenuSpec.scala:19]: Can't think of any more tests for this unless you want to add in error handling for wrong type input.  👍

File: src/test/scala/MenuSpec.scala
Comment (eloiseyaz) [inline src/test/scala/MenuSpec.scala:1]: Looks good! Now you just need to add in tests for isItemAvailable (should be pretty short) and updateMenuAfterOrder (at least 4 cases but maybe more with edge cases).

File: src/main/scala/Bill.scala (hunk 1 above)
Comment (tudorsonycx) [reply src/main/scala/Bill.scala:4]: items is a Map that maps an Item (from the menu) to an Int (its available stock). For example Map(Item(...) -> 10)
```scala
items.map({
  case (item, quantity) => item.price * quantity
})
```
is equivalent to
```scala
items.map(tuple => {
  tuple match {
    case (item, quantity) => item.price * quantity
  }
})
```
when you have a higher order function (such as map) that takes an anonymous function, if the body of the anonymous function consists of only a match statement, you can convert it to a "pattern matching anonymous function" as IntelliJ likes to call it (and in fact you'll get warned with a squiggly line to convert it)
It's just syntactic sugar, which makes it not that readable in my opinion.
the reason I used a match statement is because pattern matching allows me to destructure the tuple into its base components (an item and its stock)
I would've had to use this form if I hadn't opted for pattern matching, which is way uglier:
```scala
items.map(itemWithStock => itemWithStock._1.price * itemWithStock._2)
```
Hope that helps!

File: src/main/scala/Bill.scala (hunk 2 above)
Comment (tudorsonycx) [reply src/main/scala/Bill.scala:11]: That's a good idea, I was planning to add some more tests but I got sidetracked a bit with refactoring

File: src/main/scala/Menu.scala (hunk 3 above)
Comment (tudorsonycx) [reply src/main/scala/Menu.scala:3]: if you have a class constructor like
```scala
class MyClass private(message: String)
```
you cannot call the constructor with the `new` keyword, like
```scala
val myClass1: MyClass = new MyClass("Doesn't work")
``` 
this is because the constructor is kind of like a special method that can be made private if we wanted.
doing that forces the user to instantiate a class like this:
```scala
val myClass1: MyClass = MyClass("Works")
```
which means we have to define a companion object and implement the apply method there (which can access the private constructor because companion objects can access their classes' private fields; a case class comes with an apply method by default so we don't need to manually implement it)
in my case, it's just leftover code from something that I tried to do many commits ago and forgot to remove! thanks for pointing that out!

File: src/test/scala/CafeSpec.scala (hunk 4 above)
Comment (tudorsonycx) [reply src/test/scala/CafeSpec.scala:25]: Will do!
